Cryptokit.1.11 improvements

emillon opened the pull request:
This improves `cryptokit.1.11` on several points:
- add a missing version constraint on zarith
- fix AES-NI detection on 32 bit systems
- fix AES-NI implementation with old versions of GCC
- fall back to non-constant-time modexp (as it was in version 1.10) if only an old GMP is available

All these patches have been either picked from the upstream bug tracker, or forwarded there.

Thanks!

camelus commented:
##### :exclamation: opam-lint warnings <small>61c7e2031b0bab5658233e93a06ef0e6f02c8912</small>

* **`cryptokit.1.11`** has some warnings:

  * **warning 37**: Missing field 'dev-repo'



---


##### :white_check_mark: Installability check (5754 &rarr; 5754)

avsm commented:
cc @xavierleroy : are you ok with merging this under the cryptokit.1.11 version, or would you prefer to wait until a new release of cryptokit is ready?

xavierleroy commented:
Thanks for the notice, because the upstream bug tracker doesn't notify me when new items are posted.  

From a quick look at the proposed changes:
* add a missing version constraint on zarith -- Sure.
* fix AES-NI detection on 32 bit systems -- OK, I guess.
 * fix AES-NI implementation with old versions of GCC -- The intent is good but the implementation uses `alloca`.  No use of `alloca` will ever enter my code.
* fall back to non-constant-time modexp (as it was in version 1.10) if only an old GMP is available -- I think it's a very bad idea for a cryptographic library to silently fall back on a less secure implementation.  One of the main reasons why Cryptokit switched to GMP is to benefit from the secure modexp implementation. Can't we just demand that a sufficiently recent version of GMP is installed?

avsm commented:
Thank you for the feedback on the patches!

@emillon: I think we'd prefer consensus before merging these patches into opam-repository, especially if they override an upstream-released version number.  We can use the container distribution tests to determine a minimum version of gmp in the `conf-gmp` package, if that helps. (conf-gmp could test the required version of gmp, and then an OPAM constraint would ensure that it's available for cryptokit)

emillon commented:
@xavierleroy Thanks for the answer. I'll ping you in the future when submitting patches on the forge. Let me know if you'd prefer to discuss these patches over there.

About `alloca`, I'm not sure why I used it, the following piece of code works too (`posix_memalign` is the portable way to do that, but it allocates on the heap):

```c
__m128i unaligned_key_schedule[15+15];
__m128i *key_schedule = align16(unaligned_key_schedule);
```

About RSA, I agree about your concern with silently downgrading to a less secure implementation, though in that case I thought it could be acceptable since the previous version did not have these guarantees.

My goal is to use cryptokit on RHEL6, which is ships GMP4 (the constant-time guarantees are not important for that use case). The current situation is that unfortunately it fails at runtime. A first step is to move that failure to package installation time so that the solver can at least pick version 1.10 in such a case.

@avsm: to do that, my understanding is that I'd need to:
- add a `conf-gmp.5` package that can only install when gmp >= 5 is installed (something like `#if __GNU_MP_VERSION >= 5` or `#error` otherwise)
- add `{ >= "5"}` to the `conf-gmp` dependencies in `zarith.1.4` and `zarith.1.4.1`

If the plan sounds good to you I'll:
- change the align patch to use normal stack allocation instead of `alloca`;
- remove the RSA patch;
- open a separate PR to add the gmp package and constraints.

Then I'd like to find a solution to support 1.11 on older systems too but can't think of a way that requires explicit consent to use a non-constant time modular exponentiation.

Thanks!

avsm commented:
@emillon Hm, I need to think about how the versioning would work. I think it has to be a new package `conf-gmp5`, since it should not fail on older systems. OPAM doesnt have a way to fall back to an older revision of the package if compilation of a given version fails, so I think two package names are needed here to distinguish the particular feature.  Another option is to have `conf-gmp-<feature>` as a package that is dependent on `conf-gmp` and will fail compilation as you suggest, and depend on that from cryptokit.

> Then I'd like to find a solution to support 1.11 on older systems too but can't think of a way that requires explicit consent to use a non-constant time modular exponentiation.

Can we compile a modern GMP locally in this case?  The problem seems to older GMP packages installed in the base system, as opposed to newer GMPs not being supported on older distributions.

emillon commented:
Just updated the alignment patch and removed the RSA one.

@avsm I'll open a separate PR to add `conf-gmp-powm-sec`. Let's see if we can find a better solution.

Re. compiling a modern GMP: that works on a development machine, but makes it harder to ship binaries, so we'd end up using an old cryptokit version or maintaining a local patch to support GMP 4.

avsm commented:
thanks @emillon!
